File: mapper/AhpHostMapper.cs
// Host mapper (agent-host-protocol-core-dafny) — the host-boundary consumer shape, C# lineage.
//
// A SEPARATE compilation unit from the Dafny build tool invocation: this project
// compiles the file `dafny translate cs` emitted (build/ahp_client_cs.cs) alongside
// this hand-written harness, then DRIVES the PROVEN reducers through the host
// boundary — constructing Dafny-shaped wire values, calling a reducer, and
// reading results back out. Proves the full chain end-to-end:
//
//   Dafny model (proven)  →  dafny translate cs  →  build/ahp_client_cs.cs
//       →  separate .NET console app  →  reducer runs  →  proven outcome observed
//
// This is the C# twin of mapper/ahp-host-mapper.mjs (the JS/Node mapper). Same
// four scenarios, same assertions — ported to the real generated C# API instead
// of the real generated JS API. No stubs, no mocks: every value below is a real
// Dafny-C# datatype instance and every reducer call is the real ApplyToRoot.
//
// C# needs no facade-patching step (unlike `gen/build-js-facade.mjs`, which must
// strip the JS bundle's auto-run Main and append a CommonJS export list): the
// generated .cs already exposes real public classes/namespaces, and the
// auto-generated `__CallToMain.Main` entry point is simply never invoked because
// Mapper.csproj's <StartupObject> names OUR Main instead. See
// gen/CS-HARNESS-PATTERN.md for the reusable recipe.
using System;
using System.Numerics;

namespace Mapper
{
    public static class Program
    {
        static Dafny.ISequence<Dafny.Rune> Str(string s) => Dafny.Sequence<Dafny.Rune>.UnicodeFromString(s);
        static string FromStr(Dafny.ISequence<Dafny.Rune> s) => s.ToVerbatimString(false);
        static BigInteger Int(int n) => new BigInteger(n);

        static AhpSkeleton._IRootState RootInit() =>
            AhpSkeleton.RootState.create(
                Dafny.Sequence<AhpSkeleton._IAgentInfo>.FromElements(),
                AhpSkeleton.Option<BigInteger>.create_None(),
                AhpSkeleton.Option<Dafny.ISequence<ConfluxCodec._IJson>>.create_None(),
                AhpSkeleton.Option<AhpSkeleton._IRootConfig>.create_None());

        static AhpSkeleton._IRootState ApplyRoot(AhpSkeleton._IRootState state, AhpSkeleton._IRootAction action) =>
            AhpSkeleton.__default.ApplyToRoot(state, action, Int(0)).dtor__0;

        static int failures = 0;
        static void Check(string name, bool cond)
        {
            Console.WriteLine((cond ? "PASS" : "FAIL") + "  " + name);
            if (!cond) failures++;
        }

        public static int Main(string[] args)
        {
            // scenario 1: activeSessionsChanged sets the count (proven Applied path, observed via mapper)
            var s = ApplyRoot(RootInit(), AhpSkeleton.RootAction.create_RootActiveSessionsChanged(Int(5)));
            Check("activeSessionsChanged -> activeSessions = Some(5)",
                s.dtor_activeSessions.is_Some && s.dtor_activeSessions.dtor_value == 5);

            // scenario 2: agentsChanged replaces the agent list
            var agent = AhpSkeleton.AgentInfo.create(
                Str("copilot"), Str("Copilot"), Str("AI"),
                Dafny.Sequence<Dafny.ISequence<Dafny.Rune>>.FromElements());
            var s2 = ApplyRoot(RootInit(),
                AhpSkeleton.RootAction.create_RootAgentsChanged(Dafny.Sequence<AhpSkeleton._IAgentInfo>.FromElements(agent)));
            Check("agentsChanged -> one agent, provider \"copilot\"",
                s2.dtor_agents.Count == 1 && FromStr(s2.dtor_agents.Select(0).dtor_provider) == "copilot");

            // scenario 3: an unknown action is a no-op (loud-negative — reducer returns NoOp, state unchanged)
            var unknownRes = AhpSkeleton.__default.ApplyToRoot(
                RootInit(), AhpSkeleton.RootAction.create_RootUnknown(ConfluxCodec.Json.create_JNull()), Int(0));
            Check("unknown action -> NoOp outcome + state unchanged",
                unknownRes.dtor__1.is_NoOp && object.Equals(unknownRes.dtor__0, RootInit()));

            // scenario 4: configChanged with no config present -> no-op (precondition guard, via mapper)
            var cfgRes = AhpSkeleton.__default.ApplyToRoot(
                RootInit(),
                AhpSkeleton.RootAction.create_RootConfigChanged(
                    Dafny.Map<Dafny.ISequence<Dafny.Rune>, ConfluxCodec._IJson>.Empty, false),
                Int(0));
            Check("configChanged with no config -> NoOp (precondition guard)", cfgRes.dtor__1.is_NoOp);

            if (failures == 0)
            {
                Console.WriteLine();
                Console.WriteLine("MAPPER OK — separate .NET process drove the extracted+proven Dafny reducers to their proven outcomes");
                return 0;
            }
            else
            {
                Console.WriteLine();
                Console.WriteLine($"MAPPER FAILED — {failures} check(s) failed");
                return 1;
            }
        }
    }
}

File: mapper/ahp-host-mapper.mjs
#!/usr/bin/env node
// Host mapper (agent-host-protocol-core-dafny) — the host-boundary consumer shape.
//
// A SEPARATE process from the Dafny build: it `require`s the generated facade
// (build/ahp_client_facade.cjs) and drives the PROVEN reducers through the host
// boundary — constructing Dafny-shaped wire values, calling a reducer, and
// reading results back out. Proves the full chain end-to-end:
//
//   Dafny model (proven)  →  dafny translate js  →  facade (module.exports)
//       →  separate node consumer  →  reducer runs  →  proven outcome observed
//
// This provider-side mapper drives the pure-AHP `root` channel. (The OpenAgency
// extension repo has its own mapper driving the OA channels.)
import { createRequire } from 'node:module';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const require = createRequire(import.meta.url);
const root = dirname(dirname(fileURLToPath(import.meta.url)));
const { _dafny, BigNumber, AhpSkeleton, ConfluxCodec } = require(join(root, 'build', 'ahp_client_facade.cjs'));

const str = (s) => _dafny.Seq.UnicodeFromString(s);
const fromStr = (d) => d.toVerbatimString(false);
const int = (n) => new BigNumber(n);
const None = () => AhpSkeleton.Option.create_None();
const applyRoot = (state, action) => AhpSkeleton.__default.ApplyToRoot(state, action, int(0))[0];
const rootInit = () => AhpSkeleton.RootState.create_RootState(_dafny.Seq.of(), None(), None(), None());

let failures = 0;
const check = (name, cond) => { console.log(`${cond ? 'PASS' : 'FAIL'}  ${name}`); if (!cond) failures++; };

// scenario 1: activeSessionsChanged sets the count (proven Applied path, observed via mapper)
let s = applyRoot(rootInit(), AhpSkeleton.RootAction.create_RootActiveSessionsChanged(int(5)));
check('activeSessionsChanged → activeSessions = Some(5)',
  s.dtor_activeSessions.is_Some && s.dtor_activeSessions.dtor_value.toNumber() === 5);

// scenario 2: agentsChanged replaces the agent list
const agent = AhpSkeleton.AgentInfo.create_AgentInfo(str('copilot'), str('Copilot'), str('AI'), _dafny.Seq.of());
const s2 = applyRoot(rootInit(), AhpSkeleton.RootAction.create_RootAgentsChanged(_dafny.Seq.of(agent)));
check('agentsChanged → one agent, provider "copilot"',
  s2.dtor_agents.length === 1 && fromStr(s2.dtor_agents[0].dtor_provider) === 'copilot');

// scenario 3: an unknown action is a no-op (loud-negative — reducer returns NoOp, state unchanged)
const unknownRes = AhpSkeleton.__default.ApplyToRoot(rootInit(), AhpSkeleton.RootAction.create_RootUnknown(ConfluxCodec.Json.create_JNull()), int(0));
check('unknown action → NoOp outcome + state unchanged',
  unknownRes[1].is_NoOp && _dafny.areEqual(unknownRes[0], rootInit()));

// scenario 4: configChanged with no config present → no-op (precondition guard, via mapper)
const cfgRes = AhpSkeleton.__default.ApplyToRoot(rootInit(), AhpSkeleton.RootAction.create_RootConfigChanged(_dafny.Map.Empty, false), int(0));
check('configChanged with no config → NoOp (precondition guard)', cfgRes[1].is_NoOp);

console.log(failures === 0
  ? '\nMAPPER OK — separate node process drove the extracted+proven Dafny reducers to their proven outcomes'
  : `\nMAPPER FAILED — ${failures} check(s) failed`);
process.exit(failures === 0 ? 0 : 1);
